### PR TITLE
Refactoring tests to support both external and loopback servers

### LINKED ISSUE #73849: Add tests with LoopbackServer for web sockets
We should increase test scenarios for web sockets that are running with LoopbackServer -- first of all, for web sockets over HTTP/2 and then for HTTP/1.1.

Now many tests are running only in outerloop, which runs on request and not in every CI. This option is unavailable for web sockets over HTTP/2 because the external server has not yet supported it.

To achieve more test coverage, Http2LoopbackServer should be adapted to work with web sockets i.e., it should have a special stream wrapper to decode received data easily.

### PR DESCRIPTION
**Pull request by greenEkatherine**
Fixes #73849 

I did an example on one test `SendReceive_Concurrent_Success_Base` how I can run an existing scenarios on both external and loopback servers with different versions, security settings and attributes (like running on browser or outerloop).

Please share your thoughts about the implementation because I would like to make changes before I apply it for the remaining scenarios.

### CODE REVIEW

**Comment by greenEkatherine** [comment]
@MihaZupan @CarnaViire could you please have a look on the proposed changes? I would like to apply feedback earlier because after I add changes to other tests it will require a lot of copy paste.

**Comment by MihaZupan** [comment]
I think the pattern we use in Http tests is cleaner.
See [`HttpClientHandlerTestBase.cs`](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTestBase.cs), [`HttpClientHandlerTestBase.SocketsHttpHandler.cs`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs), [`GenericLoopbackServer.cs`](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/System/Net/Http/GenericLoopbackServer.cs).

We could add something like the following to `ClientWebSocketTestBase`:
```c#
protected virtual Version UseVersion => HttpVersion.Version11;
protected virtual bool UseRemoteServer => false;

protected Task CreateEchoServerAsync(Func<Uri, Task> clientFunc)
{
    using var server = CreateServer(UseVersion, UseRemoteServer, ...);
    await clientFunc(server.Uri);
}

protected Task<ClientWebSocket> GetConnectedWebSocket(Uri serverUri) =>
    GetConnectedWebSocket(serverUri, TimeOutMilliseconds, _output, UseVersion));
```

This can then be used in tests like
```c#
public async Task MyTest()
{
    await CreateEchoServerAsync(async uri =>
    {
        using ClientWebSocket cws = await GetConnectedWebSocket(uri);
        await cws.DoStuffAsync();
    });
}
```

This has a few benefits:
- Tests are easier to read - you don't deal with client and server tasks manually
- Tests don't have to deal with version / type of server directly
- Avoids adding `[Theory]` combinations to every test for remote/version
  - Different combinations are defined at the test class level like we have for custom invoker vs HttpClient right now
- Makes managing the lifetime of the server trivial since you never directly hand it out
  - Currently you are returning the server uri and task, but the server itself isn't getting cleaned up

File: src/libraries/System.Net.WebSockets.Client/tests/ClientWebSocketTestBase.cs [hunk 1]
```diff
@@ -141,14 +145,50 @@ protected static async Task<WebSocketReceiveResult> ReceiveEntireMessageAsync(We
             return null;
         }
 
+        protected Task CreateEchoServerAsync(Task clientTask, Task serverTask)
```
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/tests/ClientWebSocketTestBase.cs:148]: This is a bit confusing now as it isn't creating anything, it's just awaiting the provided tasks.

Right now you have tests doing
```c#
(uri, serverTask) = GetServer();

await CreateEchoServerAsync(Task.Run(async () =>
{
    using ClientWebSocket cws = await GetConnectedWebSocket(uri, TimeOutMilliseconds, _output);
    // client test stuff
}, serverTask);
```

This still has the problem regarding the lifetime of the server.
> Makes managing the lifetime of the server trivial since you never directly hand it out
Currently you are returning the server uri and task, but the server itself isn't getting cleaned up

---

Have you considered making this instead be
```c#
protected Task CreateEchoServerAsync(Func<Uri, Task> clientFunc)
```
where the implementation would effectively do
```c#
await clientFunc(RemoteServerUri);
```
for remote servers, or
```c#
using var server = WebSocketHelper.GetEchoLoopbackServer(...);
await clientFunc(server.Uri);
```
for loopback ones.

The tests can then also be simplified as they don't care about the server details
```c#
await CreateEchoServerAsync(async uri =>
{
    using ClientWebSocket cws = await GetConnectedWebSocket(uri);
    // client test stuff
});
```

File: src/libraries/System.Net.WebSockets.Client/tests/ClientWebSocketOptionsTests.cs [hunk 2]
```diff
@@ -50,7 +50,7 @@ public static void Proxy_Roundtrips()
         [ConditionalTheory(nameof(WebSocketsSupported)), MemberData(nameof(EchoServers))]
         public async Task Proxy_SetNull_ConnectsSuccessfully(Uri server)
         {
-            for (int i = 0; i < 3; i++) // Connect and disconnect multiple times to exercise shared handler on netcoreapp
+            for (int i = 0; i < 1; i++) // Connect and disconnect multiple times to exercise shared handler on netcoreapp
```
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/tests/ClientWebSocketOptionsTests.cs:53]: Was this change intentional?

File: src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 3]
```diff
@@ -12,36 +12,130 @@
...
+        public InvokerMemorySendReceiveLocalSslTest(ITestOutputHelper output) : base(output) { }
+        protected override bool UseSsl => true;
+    }
+
+    // External server -- HTTP/1.1 only----
+    // TODO: add HTTP/2 when https://corefx-net-http2.azurewebsites.net will support extended connect
+
+    [OuterLoop("Uses external servers", typeof(PlatformDetection), nameof(PlatformDetection.LocalEchoServerIsNotAvailable))]
+    [SkipOnPlatform(TestPlatforms.Browser, "Custom invoker is ignored on Browser")]
+    public sealed class InvokerMemorySendReceiveOuterSslTest : InvokerMemorySendReceiveOuterTest
```
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs:39]: Nit: `Outer` => `Remote`

File: src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 4]
```diff
@@ -419,6 +636,38 @@ public async Task SendReceive_Concurrent_Success(Uri server)
             }
         }
 
+        [ConditionalFact(nameof(WebSocketsSupported))]
+        public async Task SendReceive_Concurrent_Success_Base()
+        {
+
+            await CreateEchoServerAsync(async uri =>
+            {
+                using (ClientWebSocket cws = await GetConnectedWebSocket(uri, TimeOutMilliseconds, _output))
```
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs:645]:
```suggestion
                using ClientWebSocket cws = await GetConnectedWebSocket(uri, TimeOutMilliseconds, _output);
```
Lets you remove a set of braces here

File: src/libraries/System.Net.WebSockets.Client/tests/ClientWebSocketTestBase.cs [hunk 5]
```diff
@@ -141,14 +145,45 @@ protected static async Task<WebSocketReceiveResult> ReceiveEntireMessageAsync(We
             return null;
         }
 
+        protected Task CreateEchoServerAsync(Func<Uri, Task> clientFunc)
+        {
+            if (UseRemoteServer)
+            {
+                if (UseVersion.Major == 2)
+                {
+                    return null;
```
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/tests/ClientWebSocketTestBase.cs:154]:
```suggestion
                    throw new NotSupportedException("Remote servers don't support WebSockets over HTTP/2 yet");
```

File: src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs [hunk 6]
```diff
@@ -83,6 +91,18 @@ public static class WebSocketHelper
...
+                if (invoker == null)
+                {
+                    cws.Options.ClientCertificates.Add(Test.Common.Configuration.Certificates.GetClientCertificate());
+                    cws.Options.RemoteCertificateValidationCallback = delegate { return true; };
+                }
+
+                if (!PlatformDetection.IsBrowser)
+                {
+                    cws.Options.HttpVersion = version;
+                    cws.Options.HttpVersionPolicy = HttpVersionPolicy.RequestVersionExact;
```
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:102]: Are our automatic version upgrade/downgrade tests not using this helper?

File: src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs [hunk 7]
```diff
@@ -125,6 +145,98 @@ public static async Task<T> Retry<T>(ITestOutputHelper output, Func<Task<T>> fun
...
         }
 
+        public static Task GetEchoHttp2LoopbackServer(Func<Uri, Task> clientFunc)
+        {
+            return GetEchoHttp2LoopbackServer(clientFunc, new Http2Options());
+        }
+
+        public static Task GetEchoHttp2LoopbackServer(Func<Uri, Task> clientFunc, Http2Options options)
+        {
+            Http2LoopbackServer server = Http2LoopbackServer.CreateServer(options);
```
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:154]:
```suggestion
        public static async Task GetEchoHttp2LoopbackServer(Func<Uri, Task> clientFunc, Http2Options options)
        {
            using Http2LoopbackServer server = Http2LoopbackServer.CreateServer(options);
```

File: src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs [hunk 8]
```diff
@@ -125,6 +145,98 @@ public static async Task<T> Retry<T>(ITestOutputHelper output, Func<Task<T>> fun
...
+                        continue;
+                    }
+
+                    int offset = result.Count;
+                    await websocket.SendAsync(new ArraySegment<byte>(buffer, 0, offset), result.MessageType, true, CancellationToken.None);
+                }
+                await connection.DisposeAsync();
+            });
+
+            return new Task[] { serverTask, clientFunc(server.Address) }.WhenAllOrAnyFailed(TimeOutMilliseconds * 2);
```
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:195]:
```suggestion
            await new Task[] { serverTask, clientFunc(server.Address) }.WhenAllOrAnyFailed(TimeOutMilliseconds * 2);
```

File: src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs [hunk 9]
```diff
@@ -125,6 +144,98 @@ public static async Task<T> Retry<T>(ITestOutputHelper output, Func<Task<T>> fun
...
+                            await websocket.CloseAsync(WebSocketCloseStatus.Empty, null, CancellationToken.None);
+                        }
+                        else
+                        {
+                            WebSocketCloseStatus closeStatus = result.CloseStatus.GetValueOrDefault();
+                            await websocket.CloseAsync(
+                                closeStatus,
+                                result.CloseStatusDescription,
+                                CancellationToken.None);
+                        }
```
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:184]:
```suggestion
                        await websocket.CloseAsync(
                            result.CloseStatus ?? WebSocketCloseStatus.Empty,
                            result.CloseStatusDescription,
                            CancellationToken.None);
```

File: src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs [hunk 10]
```diff
@@ -125,6 +144,98 @@ public static async Task<T> Retry<T>(ITestOutputHelper output, Func<Task<T>> fun
...
+
+                        continue;
+                    }
+
+                    int offset = result.Count;
+                    await websocket.SendAsync(new ArraySegment<byte>(buffer, 0, offset), result.MessageType, true, CancellationToken.None);
```
Comment (MihaZupan) [inline src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:190]: The receive above may not get the complete message in a single chunk.
This should likely be passing `result.EndOfMessage` instead.

**Review by MihaZupan — COMMENTED** [review]
Nice, this looks much more familiar to the Http tests

File: src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs (hunk 6 above)
Comment (greenEkatherine) [reply src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:102]: They are not

File: src/libraries/System.Net.WebSockets.Client/tests/System.Net.WebSockets.Client.Tests.csproj [hunk 11]
```diff
@@ -71,4 +71,8 @@
   <ItemGroup>
     <PackageReference Include="System.Net.TestData" Version="$(SystemNetTestDataVersion)" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
```
Comment (CarnaViire) [inline src/libraries/System.Net.WebSockets.Client/tests/System.Net.WebSockets.Client.Tests.csproj:74]: Is this needed? the project doesn't seem to be built for .NET Framework, unless I'm missing something 🤔
https://github.com/dotnet/runtime/blob/9ea5d2d6cdc536fbe1dacc477ef2b2dea2b4845c/src/libraries/System.Net.WebSockets.Client/tests/System.Net.WebSockets.Client.Tests.csproj#L4

File: src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 12]
```diff
@@ -12,36 +12,130 @@
...
         protected override bool UseCustomInvoker => true;
     }
 
-    public sealed class HttpClientMemorySendReceiveTest : MemorySendReceiveTest
+    //HttpClient--------
+
+    // Local----
+
+    [SkipOnPlatform(TestPlatforms.Browser, "Custom invoker is ignored on Browser")]
+    public class HttpClientMemorySendReceiveLocalTest : HttpClientMemorySendReceiveTest
```
Comment (CarnaViire) [inline src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs:66]: If we are not going to update all SendReceive tests to new format in this PR, then I'm concerned that all old-format tests would run duplicatively multiple times for each of the child classes. This seems like excessive use of CI time without adding much value...

**Review by CarnaViire — CHANGES_REQUESTED** [review]
There seem to be WebSocket test failures, could you please check them?

**Comment by greenEkatherine** [comment]
> There seem to be WebSocket test failures, could you please check them?

For now there failed one unrelated build and local http2 web socket tests on Libraries Test Run release coreclr windows x86 Release. I see the intermittent failures indicating that web socket is in closed state.

File: src/libraries/System.Net.WebSockets.Client/tests/System.Net.WebSockets.Client.Tests.csproj (hunk 11 above)
Comment (greenEkatherine) [reply src/libraries/System.Net.WebSockets.Client/tests/System.Net.WebSockets.Client.Tests.csproj]: Removed condition

File: src/libraries/System.Net.WebSockets.Client/tests/wasm/System.Net.WebSockets.Client.Wasm.Tests.csproj [hunk 13]
```diff
@@ -20,30 +20,26 @@
...
-             Link="Common\System\Net\Http\LoopbackProxyServer.cs" />
-    <Compile Include="$(CommonTestPath)System\Net\Http\LoopbackServer.cs"
-             Link="Common\System\Net\Http\LoopbackServer.cs" />
-    <Compile Include="$(CommonTestPath)System\Net\Http\GenericLoopbackServer.cs"
-             Link="Common\System\Net\Http\GenericLoopbackServer.cs" />
-    <Compile Include="$(CommonTestPath)System\Security\Cryptography\PlatformSupport.cs"
-             Link="CommonTest\System\Security\Cryptography\PlatformSupport.cs" />
-    <Compile Include="$(CommonTestPath)System\Threading\Tasks\TaskTimeoutExtensions.cs"
-             Link="Common\System\Threading\Tasks\TaskTimeoutExtensions.cs" />
+    <Compile Include="$(CommonTestPath)System\Net\Capability.Security.cs" Link="Common\System\Net\Capability.Security.cs" />
```
[the remark this answers is not in the document]
Comment (greenEkatherine) [reply src/libraries/System.Net.WebSockets.Client/tests/wasm/System.Net.WebSockets.Client.Wasm.Tests.csproj:23]: There were files added and others are auto-formatted

File: src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs [hunk 14]
```diff
@@ -419,6 +636,37 @@ public async Task SendReceive_Concurrent_Success(Uri server)
             }
         }
 
+        [ConditionalFact(nameof(WebSocketsSupported))]
+        public async Task SendReceive_Concurrent_Success_Base()
```
[the remark this answers is not in the document]
Comment (greenEkatherine) [reply src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs]: Yes

File: src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs (hunk 12 above)
Comment (greenEkatherine) [reply src/libraries/System.Net.WebSockets.Client/tests/SendReceiveTest.cs:66]: Condition added

File: src/libraries/System.Net.WebSockets.Client/tests/System.Net.WebSockets.Client.Tests.csproj (hunk 11 above)
Comment (greenEkatherine) [reply src/libraries/System.Net.WebSockets.Client/tests/System.Net.WebSockets.Client.Tests.csproj:74]: It breaks builds, reverted

**Comment by CarnaViire** [comment]
I will take over this PR, approximately in a month. Marking as draft for now.
